### PR TITLE
Rework the container image build actions

### PR DESCRIPTION
**Pull request by renchap**
Those actions are now using a re-usable workflow: `build-container-image.yml`.

Several container builds are now triggered:

- Release builds, for every git tag. Those are pushed to both repositories
- Nightly build, every night, pushing to `nightly` and `edge` tags (which is no longer for every merge on `main`)
- Test builds, when the Dockerfile or the container-related workflows are changed. Those images are not pushed anywhere
- On-demand builds for specific PRs. If you label a PR with the `build-image` label, then a Docker build will run for every push on this PR, and the resulting image will be pushed to the Github Registry for testing

Breaking change: the `edge` tag is no longer tracking `main`, but is similar to `nightly`, for compatibility, but it should not be used (please pin to specific versions).

### CODE REVIEW

**Comment by renchap** [comment]
I integrated the native ARM64 builder support from #25855

The container image test build is only using AMD64, so it runs natively and is still performant, while not using the remote ARM64 builder to avoid having arbitrary code run on it.

The other builds use the native remote builder if they build for ARM64.

**Review by Gargron — APPROVED** [review]

File: .github/workflows/build-nightly.yml [hunk 1]
```diff
@@ -1,60 +1,39 @@
-name: Build nightly container image
+name: Build container release images
```
Comment (ClearlyClaire) [inline .github/workflows/build-nightly.yml:1]: Why the name change?

File: .github/workflows/build-push-pr.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+          echo mastodon_version_suffix=+pr-${{ github.event.pull_request.number }}-$(git rev-parse --short HEAD) >> $GITHUB_OUTPUT
+    outputs:
+      suffix: ${{ steps.version_vars.outputs.mastodon_version_suffix }}
+
+  build-image:
+    needs: compute-suffix
+    uses: ./.github/workflows/build-container-image.yml
+    with:
+      platforms: linux/amd64,linux/arm64
+      use_native_arm64_builder: true
```
Comment (ClearlyClaire) [inline .github/workflows/build-push-pr.yml:26]: Do we want native ARM64 builds for these PRs?

**Review by ClearlyClaire — COMMENTED** [review]
> On-demand builds for specific PRs. If you label a PR with the build-image label, then a Docker build will run for every push on this PR, and the resulting image will be pushed to the Github Registry for testing

How does this work? The file contains:
```yaml
on:
  pull_request:
    types: [labeled, synchronize, reopened, ready_for_review, opened]
```
but I cannot see any reference to a specific label

File: .github/workflows/build-push-pr.yml (hunk 2 above)
Comment (renchap) [reply .github/workflows/build-push-pr.yml:26]: I added it as it should now be fast with the native builder, and it allows to manually trigger a full Docker build for any PR, which might be useful to test if it really works when doing some build changes

**Comment by renchap** [comment]
> How does this work? The file contains:
> 
> ```yaml
> on:
>   pull_request:
>     types: [labeled, synchronize, reopened, ready_for_review, opened]
> ```
> 
> but I cannot see any reference to a specific label

The label is checked in the job directly, as you cannot filter in trigger:

https://github.com/mastodon/mastodon/blob/8ff5aaf69796e51a19f89872bbdb5ffd14201002/.github/workflows/build-push-pr.yml#L13

File: .github/workflows/build-push-pr.yml (hunk 2 above)
Comment (ClearlyClaire) [reply .github/workflows/build-push-pr.yml:26]: Sounds good, although that means we need to restrict who can set those labels.

Comment (renchap) [reply .github/workflows/build-push-pr.yml:26]: Labels can be set by people in the organisation, so triage team + devs. I think it mitigates sufficiently the issue.

**Review by ClearlyClaire — APPROVED** [review]

**Review by timetinytim — APPROVED** [review]

**Pull request merged**